test/functional/client_side_encryption_prose_tests.js

```diff
@@ -0,0 +1,974 @@
+              encrypted_placeholder: {
+                encrypt: {
+                  keyId: '/placeholder',
+                  bsonType: 'string',
+                  algorithm: 'AEAD_AES_256_CBC_HMAC_SHA_512-Random'
+                }
+              }
+            }
+          }
+        };
+
+        return (
+          Promise.resolve()
+            .then(() => this.client.connect())
+            // #. Using ``client``, drop the collections ``admin.datakeys`` and ``db.coll``.
+            .then(() => this.client.db(dataDbName).dropCollection(dataCollName))
+            .catch(noop)
+            .then(() => this.client.db(keyVaultDbName).dropCollection(keyVaultCollName))
+            .catch(noop)
+            // #. Create the following:
+            //   - A MongoClient configured with auto encryption (referred to as ``client_encrypted``)
+            //   - A ``ClientEncryption`` object (referred to as ``client_encryption``)
+            //   Configure both objects with ``aws`` and the ``local`` KMS providers as follows:
+            //   .. code:: javascript
+            //       {
+            //           "aws": { <AWS credentials> },
+            //           "local": { "key": <base64 decoding of LOCAL_MASTERKEY> }
+            //       }
+            //   Configure both objects with ``keyVaultNamespace`` set to ``admin.datakeys``.
+            //   Configure the ``MongoClient`` with the following ``schema_map``:
+            //   .. code:: javascript
+            //       {
+            //         "db.coll": {
+            //           "bsonType": "object",
+            //           "properties": {
+            //             "encrypted_placeholder": {
+            //               "encrypt": {
+            //                 "keyId": "/placeholder",
+            //                 "bsonType": "string",
+            //                 "algorithm": "AEAD_AES_256_CBC_HMAC_SHA_512-Random"
+            //               }
+            //             }
+            //           }
+            //         }
+            //       }
+            //   Configure ``client_encryption`` with the ``keyVaultClient`` of the previously created ``client``.
+            .then(() => {
+              this.clientEncryption = new mongodbClientEncryption.ClientEncryption(this.client, {
+                kmsProviders,
+                keyVaultNamespace
+              });
+            })
+            .then(() => {
+              this.clientEncrypted = this.configuration.newClient(
+                {},
+                {
+                  useNewUrlParser: true,
+                  useUnifiedTopology: true,
+                  autoEncryption: {
+                    keyVaultNamespace,
+                    kmsProviders,
+                    schemaMap
+                  }
+                }
+              );
+              return this.clientEncrypted.connect();
+            })
+        );
+      });
+
+      afterEach(function() {
+        this.commandStartedEvents = [];
```

Comment:
I noticed you're adding to the local test context (this) rather than using a floating variable lexically scoped to the `describe` like you normally do, any particular reason?

Reply:
Personal preference. I like not having empty `let`s sitting at the top of the scope. I definitely read once that it was encouraged behavior, but I cannot find the source.